Client/Frontend/Sync/BraveCore/BraveCoreDataImportExportUtility.swift

```diff
@@ -0,0 +1,475 @@
+                    try file.write(currentIndentation)
+                    try file.write(FolderWriter.kFolderChildrenEnd)
+                    try file.write(BraveCoreDataImportExportUtility.kNewline)
+                }
+            } else {
+                try file.write(currentIndentation)
+                
+                //Start Bookmark
+                try file.write(BookmarkWriter.kBookmarkStart)
+                
+                //URL
+                try file.write(bookmark.url?.escapeAsHTMLQuotes() ?? "")
+                
+                //Created Date
+                try file.write("\(Int(bookmark.created?.timeIntervalSince1970 ?? Date().timeIntervalSince1970))")
+                
+                //ICON (OPTIONAL) - 16x16 PNG
+                if let iconUrl = bookmark.domain?.favicon?.url,
+                   let url = URL(string: iconUrl),
+                   let iconData = try? Data(contentsOf: url),
+                   let image = UIImage(data: iconData, scale: UIScreen.main.scale)?.scale(toSize: CGSize(width: 16.0, height: 16.0)),
+                   let pngData = image.pngData() {
+                    try file.write(BookmarkWriter.kIcon)
+                    try file.write("data:image/png;base64,\(pngData.base64EncodedString)".escapeAsHTMLQuotes())
+                }
+                
+                //End Attributes
+                try file.write(BookmarkWriter.kBookmarkAttributeEnd)
+                
+                //Title
+                try file.write((bookmark.title ?? "(No Title)").escapeForHTML())
+                
+                //End Bookmark
+                try file.write(BookmarkWriter.kBookmarkEnd)
+                try file.write(BraveCoreDataImportExportUtility.kNewline)
+            }
+        }
+    }
+    
+    private func incrementIndent() {
+        currentIndentation += String(repeating: " ", count: BraveCoreDataImportExportUtility.kIndentSize)
+    }
+    
+    private func decrementIndent() {
+        if currentIndentation.count >= BraveCoreDataImportExportUtility.kIndentSize {
+            currentIndentation.removeLast(BraveCoreDataImportExportUtility.kIndentSize)
+        }
+    }
+    
+    private func rethrow(_ closure: @autoclosure () throws -> Void, error: ParsingError) rethrows {
+        do {
+            try closure()
+        } catch {
+            throw error
+        }
+    }
+    
+    // File header.
+    private static let kHeader =
+        "<!DOCTYPE NETSCAPE-Bookmark-file-1>\r\n" +
+        "<!-- This is an automatically generated file.\r\n" +
+        "     It will be read and overwritten.\r\n" +
+        "     DO NOT EDIT! -->\r\n" +
+        "<META HTTP-EQUIV=\"Content-Type\"" +
+        " CONTENT=\"text/html; charset=UTF-8\">\r\n" +
+        "<TITLE>Bookmarks</TITLE>\r\n" +
+        "<H1>Bookmarks</H1>\r\n" +
+        "<DL><p>\r\n"
+
+    // Newline separator.
+    private static let kNewline = "\r\n"
+
+    private struct BookmarkWriter {
```

Comment:
is there an XSS risk here? for instance if someone's bookmark title is `\"</a><script>alert(1)</script>`

Reply:
i guess not because all the external input here is HTML-escaped

Reply:
just to be paranoid, i think we should HTML escape every injected field including date and icon.